### PR TITLE
Refactor ol style creation

### PR DESCRIPTION
**Pull request by ajkopinga**
While working on optimizing GraphicStroke performance, I also wanted to enable GraphicStroke for PolygonSymbolizers and GraphicStroke with ExternalGraphics. This proved to be hard to do without refactoring almost all of the ol style creation code.

This pull request:
* Improves GraphicStroke performance by caching the temporary OL point style instance.
* Enables GraphicStroke inside a PolygonSymbolizer Stroke element.
* Enables ExternalGraphic inside GraphicStroke.
* Now correctly handles SLD defaults for missing elements/style properties.
* Fixes a bug where ExternalGraphic did not work when there's more than one symbolizer inside a Rule.
* Handles simple Fill/Stroke creation in one place, instead of separately for each place where point/line/polygon styles are created.
* Refactors the graphic style invalidation code, making it more maintainable.

### CODE REVIEW

File: src/imageCache.js [hunk 1]
```diff
@@ -170,46 +213,34 @@ export function processExternalGraphicSymbolizers(
   for (let k = 0; k < rules.length; k += 1) {
     const rule = rules[k];
 
-    let symbolizer;
-    let exgraphic;
+    const allSymbolizers = getAllSymbolizers(rule);
+    allSymbolizers.forEach(symbolizer => {
+      externalGraphicPaths.forEach(path => {
```
Comment (allartk) [inline src/imageCache.js:218]: Many nested closures does not increase readability. Maybe some functions can be moved outside the loop.

File: src/imageCache.js [hunk 2]
```diff
@@ -170,46 +213,34 @@ export function processExternalGraphicSymbolizers(
...
-    ) {
-      updateExternalGraphicRule(rule, imageUrl, imageLoadState[imageUrl]);
-    }
+        // When an external graphic has been found inside a symbolizer,
+        // either start loading the image, or check if the load state of the image has changed.
+        const imageUrl = exgraphic.onlineresource;
+        if (!(imageUrl in imageLoadState)) {
+          // Start loading the image and set image load state on the symbolizer.
+          imageLoadState[imageUrl] = IMAGE_LOADING;
+          loadExternalGraphic(
```
Comment (allartk) [inline src/imageCache.js:230]: It seems you choose to set the imageloadstate inside the sld symbolizer object. Can this concern be moved into a separate object?

File: src/imageCache.js [hunk 3]
```diff
@@ -170,46 +213,34 @@ export function processExternalGraphicSymbolizers(
   for (let k = 0; k < rules.length; k += 1) {
     const rule = rules[k];
 
-    let symbolizer;
-    let exgraphic;
+    const allSymbolizers = getAllSymbolizers(rule);
```
[the remark this answers is not in the document]
Comment (ajkopinga) [reply src/imageCache.js:216]: Agreed. I will change this.

File: src/imageCache.js (hunk 1 above)
Comment (ajkopinga) [reply src/imageCache.js:218]: I'm afraid this degree of nesting is necessary. For each rule, all symbolizers have to be examined. For each symbolizer, all possible image locations (externalGraphicPaths) have to be checked to see if the symbolizer contains an image, and if that image's load state has changed.

File: src/imageCache.js (hunk 2 above)
Comment (ajkopinga) [reply src/imageCache.js:230]: I will move checking image load states to getters/setters, just like for the cached image, which also has getter/setter for a global cache. I think that will improve the clarity of what's happening here. After doing that, I think I can stop storing the load state flag inside the symbolizer.

**Pull request merged**